### PR TITLE
Fix deprecation: preview=in

### PR DESCRIPTION
**Pull request by Inkrementator**
Please check lines that contain "REVIEW" in a comment, they need verification and I suspect you can do it faster than me.

Unit tests fail to compile. The offending line looked intentional to me, so I left it in.

### CODE REVIEW

**Comment by Inkrementator** [comment]
Since "in" seems to pass the parameters by value, which is not what you did beforehand, declaring them all "const ref" might be more desirable. I don't know if this matters

**Comment by Inkrementator** [comment]
Relevant dmd release for the deprecation:

https://dlang.org/changelog/2.104.0.html

**Review by schveiguy — CHANGES_REQUESTED** [review]
So let's just do a change without preview=in as an always-on flag. Maybe put it inside a configuration for testing. I can live with passing socket addresses by value if the user doesn't enable preview-in.

File: src/std/io/net/addr.d [hunk 1]
```diff
@@ -848,17 +848,17 @@ else version (Windows)
 private:
 
 // Won't actually fail with proper buffer size and typed addr, so it'll never set errno and is actually pure
-alias pure_inet_ntop = extern (System) const(char)* function(int, in void*, char*, socklen_t) pure nothrow @nogc;
+alias pure_inet_ntop = extern (System) const(char)* function(int,scope const(void)*, char*, socklen_t) pure nothrow @nogc; //REVIEW in void* = scope const void*?
```
[the remark this answers is not in the document]
Comment (Inkrementator) [reply src/std/io/net/addr.d:851]: Error: cannot use `in` parameters with `extern(C)` functions

https://dlang.org/changelog/2.104.0.html#dmd.in-externd

[the remark this answers is not in the document]
Comment (schveiguy) [reply src/std/io/net/addr.d:851]: Oh! OK, let me take a look at the requirements.

[the remark this answers is not in the document]
Comment (schveiguy) [reply src/std/io/net/addr.d:851]: OK, yeah, this should be fine and legal.

File: src/std/io/net/addr.d [hunk 2]
```diff
@@ -848,17 +848,17 @@ else version (Windows)
...
 // Won't actually fail with proper addr family, so it'll never set errno and is actually pure
-alias pure_inet_pton = extern (System) int function(int, in char*, void*) pure nothrow @nogc;
+alias pure_inet_pton = extern (System) int function(int, scope const(char)*, void*) pure nothrow @nogc;
```
[the remark this answers is not in the document]
Comment (schveiguy) [reply src/std/io/net/addr.d:853]: As per other comment, this is fine.

File: dub.sdl [hunk 3]
```diff
@@ -6,10 +6,13 @@ license "BSL-1.0 (Boost)"
...
-dflags "-preview=in"
 configuration "library" {
 }
 configuration "dip1000" {
        dflags "-dip1000"
        versions "DIP1000"
 }
+configuration "preview_in" {
+	dflags "-preview=in"
+	versions "preview_in"
```
Comment (Inkrementator) [inline dub.sdl:17]: Does enabling the flag "-preview=in" add any version so this line is unnecessary?

File: src/std/io/exception.d [hunk 4]
```diff
@@ -73,7 +73,11 @@ unittest
     }
 
     scope e = new IOException(String("Hello, World"));
-    e.toString(&old);
+    version(preview_in)
+        assert(!__traits(compiles, e.toString(&old)));
```
Comment (schveiguy) [inline src/std/io/exception.d:77]: Hm... this is a sign something is wrong.

**Comment by Inkrementator** [comment]
I forced pushed a new version that passes unittests with SKIP_IPv6_LOOPBACK_TESTS=1 because my system doesn't support it.

File: src/std/io/exception.d (hunk 4 above)
Comment (Inkrementator) [reply src/std/io/exception.d:77]: With preview in, the compiler decides whether the type is `const scope char[]` or `const scope ref char[]`. Without it, `in` is well defined and it compiles

File: dub.sdl (hunk 3 above)
Comment (schveiguy) [reply dub.sdl:17]: No, but I don't like the idea of having a version for this. Please remove.

File: src/std/io/exception.d (hunk 4 above)
Comment (schveiguy) [reply src/std/io/exception.d:77]: OK, let's just remove this whole part (the `old` function usage). Use the `new_` thing only. Looking back at the reason this was added, it wasn't passing CI at the time, and this was needed to update the compiler (std.io used to be in dmd buildkite).

**Comment by schveiguy** [comment]
I really need to get the CI working again on this...

**Comment by Inkrementator** [comment]
Btw I think I ticked some checkbox that you can just make changes to my PR. It might be faster for you to make small changes yourself rather than commenting them, I don't know how convenient this system is.

**Comment by schveiguy** [comment]
> Btw I think I ticked some checkbox that you can just make changes to my PR. It might be faster for you to make small changes yourself rather than commenting them, I don't know how convenient this system is.

Yeah, thanks. I don't really know how to use that system easily... My understanding is I have to use the gh command line tool, or sign up for something else.

I'm OK to do this if you feel like dropping it. I actually have to get it done either way, as others have separately requested me to fix `io` for preview=in. You've helped me a lot already!

**Comment by schveiguy** [comment]
Actually, I think this is good to go. I just have to test it, or get the CI to start working...

**Review by schveiguy — APPROVED** [review]

**Comment by Inkrementator** [comment]
I'm fine helping like this, just want to make sure you don't think I'll be offended if you throw it all out and do it your own way.

**Comment by schveiguy** [comment]
hehe, thanks. Yes, I'm also a not-get-offended type of person.

File: dub.sdl [hunk 5]
```diff
@@ -12,3 +12,6 @@ configuration "dip1000" {
        dflags "-dip1000"
        versions "DIP1000"
 }
+configuration "preview_in" {
+	dflags "-preview=in"
+}
```
Comment (Geod24) [inline dub.sdl:17]: Unless this is used in the CI, this is not a good idea. The way to pass previews is to pass them from the application recipe file via the dependencies' dflags.

Comment (schveiguy) [reply dub.sdl:17]: Does setting the dflags in the main project make it to the dependencies? My point was that you can use this configuration to turn on preview=in for your dependency.

I know that setting the environment DFLAGS goes everywhere, but I thought dflags was specific to the project itself.

Comment (schveiguy) [reply dub.sdl:17]: So this leaves a lot to be desired. If I specify "dflags" in my application, I get the flag there, but not on the dependencies. I have to specify the flags in the dependency (which means a more complex dependency specification), then I get it on the dependency, but then I have to specify on *every* dependency, and their subdependencies. Plus I have to specify in the application itself.

But yes, I agree these configs aren't really proper. We need a better way to say "for all compilations, use these flags".

**Comment by schveiguy** [comment]
I tested locally, passed all existing unittests. Good enough for me.

File: dub.sdl (hunk 5 above)
Comment (Geod24) [reply dub.sdl:17]: >  but then I have to specify on every dependency, and their subdependencies. 

No, if you specify it on a dependency, it is transitive. The rest is correct. I agree the current situation is suboptimal but that was the only way we could get things to work without changing dub's behavior.

**Pull request merged**